### PR TITLE
Fix hashing for python 3.9

### PR DESCRIPTION
**Pull request by lhoestq**
In python 3.9, pickle hashes the `glob_ids` dictionary in addition to the `globs` of a function.

Therefore the test at `tests/test_fingerprint.py::RecurseDumpTest::test_recurse_dump_for_function_with_shuffled_globals` is currently failing for python 3.9

To make hashing deterministic when the globals are not in the same order, we also need to make the order of `glob_ids` deterministic.

Right now we don't have a CI to test python 3.9 but we should definitely have one. For this PR in particular I ran the tests locally using python 3.9 and they're passing now.

Fix https://github.com/huggingface/datasets/issues/4506

### CODE REVIEW

**Comment by lhoestq** [comment]
What do you think @albertvillanova ?

File: src/datasets/utils/py_utils.py [hunk 1]
```diff
@@ -818,6 +818,22 @@ def save_function(pickler, obj):
...
+            # DONE: modified here for huggingface/datasets
+            # - globs is a dictionary with keys = var names (str) and values = python objects
+            # - globs_copy is a dictionary with keys = var names (str) and values = ids of the python objects
+            # however the dictionary is not always loaded in the same order
+            # therefore we have to sort the keys to make deterministic.
+            # This is important to make `dump` deterministic.
+            # Only these line are different from the original implementation:
+            # START
+            globs_is_globs_copy = globs is globs_copy
+            globs = {k: globs[k] for k in sorted(globs.keys())}
```
Comment (albertvillanova) [inline src/datasets/utils/py_utils.py:830]:
```suggestion
            globs = dict(sorted(globs.items()))
```

File: src/datasets/utils/py_utils.py [hunk 2]
```diff
@@ -818,6 +818,22 @@ def save_function(pickler, obj):
...
+            if globs_is_globs_copy:
+                globs_copy = globs
+            elif globs_copy is not None:
+                globs_copy = {k: globs_copy[k] for k in sorted(globs_copy.keys())}
```
Comment (albertvillanova) [inline src/datasets/utils/py_utils.py:834]:
```suggestion
                globs_copy = dict(sorted(globs_copy.items()))
```

**Review by albertvillanova — APPROVED** [review]
Thank you for your foresight! Good.

I think this will be probably integrated into `dill`. ;)
cc: @anivegesana

Just some unimportant proposed refactorings below. Feel free to ignore them.

**Pull request merged**